migration/iD10/oracle/202209091013_IDEMPIERE-5410.sql

```diff
@@ -0,0 +1,140 @@
+-- IDEMPIERE-5410-kpi
+SELECT register_migration_script('202209091013_IDEMPIERE-5410.sql') FROM dual;
+
+SET SQLBLANKLINES ON
+SET DEFINE OFF
+
+-- Sep 9, 2022, 10:13:47 AM CEST
+INSERT INTO AD_Message (MsgType,MsgText,AD_Client_ID,AD_Org_ID,IsActive,Created,CreatedBy,Updated,UpdatedBy,AD_Message_ID,Value,EntityType,AD_Message_UU) VALUES ('I','<div class="kpi-title">
+			{0,number,#,###} EUR</div>
+<div style="kpi-text-align: center;">
+	<div class="kpi-sub-text">Last 365 days</div>
+</div>',0,0,'Y',TO_TIMESTAMP('2022-09-09 10:13:46','YYYY-MM-DD HH24:MI:SS'),100,TO_TIMESTAMP('2022-09-09 10:13:46','YYYY-MM-DD HH24:MI:SS'),100,200782,'KPI-AVG-1','D','51be543f-cc86-4962-bbc7-cf0afb10a0cd')
+;
+
+-- Sep 9, 2022, 10:16:28 AM CEST
+INSERT INTO AD_Message (MsgType,MsgText,AD_Client_ID,AD_Org_ID,IsActive,Created,CreatedBy,Updated,UpdatedBy,AD_Message_ID,Value,EntityType,AD_Message_UU) VALUES ('I','<div class="kpi-title">
+			{0,number,#,###.##} %</div>
+		<div class="kpi-text">Gross Profit Margin YTD</div>
+
+		<div style="kpi-text-align: center;">
+			<div class="kpi-text {2,choice,-1#kpi-data-negative|0#|0<kpi-data-positive}">
+				<div class="kpi-arrow {2,choice,-1#kpi-arrow-down|0#kpi-arrow-right|0<kpi-arrow-up}"></div>
+				{2,number,#.#} %
+		</div>
+		<div class="kpi-sub-text">vs {1,number,#,###.##} % Previous YTD</div>
+</div>',0,0,'Y',TO_TIMESTAMP('2022-09-09 10:16:28','YYYY-MM-DD HH24:MI:SS'),100,TO_TIMESTAMP('2022-09-09 10:16:28','YYYY-MM-DD HH24:MI:SS'),100,200783,'KPI-2-YTD','D','f88368c6-a767-4022-88ab-16db4de3997e')
+;
+
+-- Sep 9, 2022, 10:17:02 AM CEST
+INSERT INTO AD_StatusLine (AD_StatusLine_ID,AD_Client_ID,AD_Org_ID,AD_StatusLine_UU,Created,CreatedBy,EntityType,IsActive,Name,Updated,UpdatedBy,AD_Message_ID,SQLStatement) VALUES (200021,0,0,'63f54ea1-b43f-4f43-9af1-6d8a591015bb',TO_TIMESTAMP('2022-09-09 10:17:02','YYYY-MM-DD HH24:MI:SS'),100,'D','Y','KPI/Average Sales',TO_TIMESTAMP('2022-09-09 10:17:02','YYYY-MM-DD HH24:MI:SS'),100,200782,'SELECT AVG(TotalLines) FROM C_Order WHERE AD_CLient_ID = @#AD_Client_ID@ AND issotrx=''Y'' AND IsProposal = ''N'' AND DATE_TRUNC(''day'',dateordered) BETWEEN DATE_TRUNC(''day'',NOW() -INTERVAL ''1 year'') AND  DATE_TRUNC(''day'',NOW())')
+;
+
+-- Sep 9, 2022, 10:17:20 AM CEST
+INSERT INTO AD_StatusLine (AD_StatusLine_ID,AD_Client_ID,AD_Org_ID,AD_StatusLine_UU,Created,CreatedBy,EntityType,IsActive,Name,Updated,UpdatedBy,AD_Message_ID,SQLStatement) VALUES (200022,0,0,'53f41a48-b5fc-4016-b8b6-5c450efff3ab',TO_TIMESTAMP('2022-09-09 10:17:20','YYYY-MM-DD HH24:MI:SS'),100,'D','Y','KPI_Gross Profit Margin YTD',TO_TIMESTAMP('2022-09-09 10:17:20','YYYY-MM-DD HH24:MI:SS'),100,200783,'SELECT 
+  ROUND (SUM(((revenue-COGS)/revenue)*100) FILTER (WHERE kpidata.year = date_trunc(''year'', now())),2)::numeric as revenue_YTD ,
```

Comment:
we need to include here SQL that works both in oracle and postgresql

otherwise the demo won't work in oracle

Reply:
we forgot about oracle and kpi status line. we have no experience with oracle.
i understand this requirement, no idea for now how to make compatible. (we need understand oracle syntax)
hope i understood the problem - the kpi status lines makes data incompactible.

can be kpi definitions different in PG and oracle ? that is acceptable or we must use common/compatible  syntax ? 
n

Reply:
hi @norbertbede , preferably we should use common syntax to reduce maintenance issue. Some observation and suggestion:
- trunc can be use in place of date_trunc. trunc(timestamp, 'YEAR') and trunc(timestamp, 'MONTH')
- ::numeric, there's no common numeric type between oracle and postgresql. the solution here is either to remove the use of ::numeric and if it is a must, introduce a `@NUMERIC_DB_TYPE@` context variable and replace ::numeric with `cast(... as @NUMERIC_DB_TYPE@)`. Alternatively, I think with Oracle dialect mode, cast(... as Number) should work as the translate layer will convert Number to Numeric.
- there's no equivalent of generate_series in Oracle. While we can create similar function in Oracle, the sql syntax of using it will be slightly different from PostgreSQL.

Another option is to distribute this as postgresql only plugin and free your self from the Core constraint of having to support both Oracle and PostgreSQL.

Reply:
@hengsin is right, additionally:

* most of the time we can simply remove the ::numeric (and any ::casting) and postgresql works fine, it is required in very few cases
* I see also in the queries SUM ... FILTER - that's postgresql only syntax, and can be expressed in a generic way as SUM ... CASE WHEN ...
* generate_series has no equivalent, in the case of the query here, as it is generating 12 records it could be replaced in a generic way with a JOIN with a subquery table with 12 SELECTS joined by UNION

Reply:
Hi @CarlosRuiz-globalqss, @hengsin 

In the latest commit I removed the ::numeric casts, replaced date_trunc with trunc function and replaced FILTER with CASE WHEN statement, we would like you to review if this would be ok. 

The remaining question is that what would be the best way of replacing/removing generate_series?

Reply:
would be great use similar approach as on truncate().
as @CarlosRuiz-globalqss suggest
=> there's no equivalent of generate_series in Oracle. While we can create similar function in Oracle, the sql syntax of using it will be slightly different from PostgreSQL.
can you help with this ?